internal/controllers/configuration/object_references.go

```diff
@@ -0,0 +1,157 @@
+package configuration
+
+import (
+	"context"
+
+	corev1 "k8s.io/api/core/v1"
+	extv1beta1 "k8s.io/api/extensions/v1beta1"
+	netv1 "k8s.io/api/networking/v1"
+	netv1beta1 "k8s.io/api/networking/v1beta1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/annotations"
+	ctrlref "github.com/kong/kubernetes-ingress-controller/v2/internal/controllers/reference"
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/dataplane"
+	kongv1 "github.com/kong/kubernetes-ingress-controller/v2/pkg/apis/configuration/v1"
+	kongv1beta1 "github.com/kong/kubernetes-ingress-controller/v2/pkg/apis/configuration/v1beta1"
+)
+
+// updateReferredObjects updates reference records where the referrer is the object in parameter obj.
+// currently it only updates reference records to secrets, since we wanted to limit cache size of secrets:
+// https://github.com/Kong/kubernetes-ingress-controller/issues/2868
+func updateReferredObjects(
+	ctx context.Context, client client.Client, refIndexers ctrlref.CacheIndexers, dataplaneClient *dataplane.KongClient, obj client.Object) error {
+
+	var referredSecretNameMap = make(map[types.NamespacedName]struct{})
+	var referredSecretList []types.NamespacedName
+	switch obj := obj.(type) {
+	// functions update***ReferredSecrets first list the secrets referred by object,
+	// then call UpdateReferencesToSecret to store referrence records between the object and referred secrets,
+	// and also to remove the outdated reference records in cache where the secret is not referred by the obj specification anymore.
+	case *corev1.Service:
+		referredSecretList = listCoreV1ServiceReferredSecrets(obj)
+	case *netv1.Ingress:
+		referredSecretList = listNetV1IngressReferredSecrets(obj)
+	case *netv1beta1.Ingress:
+		referredSecretList = listNetV1beta1IngressReferredSecrets(obj)
+	case *extv1beta1.Ingress:
+		referredSecretList = listExtensionV1beta1IngressReferredSecrets(obj)
+	case *kongv1.KongPlugin:
+		referredSecretList = listKongPluginReferredSecrets(obj)
+	case *kongv1.KongClusterPlugin:
+		referredSecretList = listKongClusterPluginReferredSecrets(obj)
+	case *kongv1.KongConsumer:
+		referredSecretList = listKongConsumerReferredSecrets(obj)
+	case *kongv1beta1.TCPIngress:
+		referredSecretList = listTCPIngressReferredSecrets(obj)
+	}
+
+	for _, nsName := range referredSecretList {
+		referredSecretNameMap[nsName] = struct{}{}
+	}
```

Comment:
I wonder why do we need this as a map? I can't seem to find a place where the key lookup is being used.

I believe we can use the slice (that we already have) `referredSecretList`, no?

Reply:
I'm suspecting @randmonkey's intention was to make sure we have no duplicates here (which is fine I guess to ensure that in case some of the functions getting related secrets could return a duplicate). I'd only prefer to make this deduplication encapsulated in a tiny function accepting and returning a slice, not leaking the map outside.

Reply:
Good point!

Reply:
No, the main purpose for using map is to call `removeOutdatedReferencesToSecret` to remove outdated references. In this function, we need to list the references already saved in cache, and remove the references that are not in the k8s project. For example, An ingress `I` initially referred to secret `A`, then it is updated to refer to secret `B` and not to refer to secret `A`, we should add reference `I->A` initially, then remove `I-A` and add `I->B` after `I` is updated. So we need a map to find out what secrets are actually referred, and it would be better to pass through a map for the checking.

If you feel really uncomfortable to see maps passed between functions, I think the make map procedure could be put in `UpdateReferencesToSecret` function.

Reply:
I'm OK with leaving this as is but I thought that since we don't use the lookup (right?) we might swap it for a slice but since the deduplication is a worthy use case, I'm OK with the current state of this.

Reply:
> I'm OK with leaving this as is but I thought that since we don't use the lookup (right?) we might swap it for a slice but since the deduplication is a worthy use case, I'm OK with the current state of this.

lookup is used, but not directly in `UpdateReferencesToSecret`.